Fix percentile inconsistencies

Oisin-M opened the pull request:
- [x] Closes #11336 #11989
- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

This PR makes some small changes to the percentile function.
- simplifies logic slightly and adds minor optimisations
- fixes issue with inconsistent output shapes (does a `.reshape(())` if required)
- corrects some erroneous docstrings

jacobtomlinson reviewed (commented):
Overall I think this looks great.

Do you think you could add a little more testing? For example in #11989 there is an example of some assertions that were failing. Could you update the tests to demonstrate they now pass?

Oisin-M commented:
Hi @jacobtomlinson, I already updated the `test_percentiles_with_scaler_percentile` to test for the issue in #11989 but I went and added some further tests now too. We only get approximate matches with numpy's percentiles by algorithmic design so I've added a very lenient relative tolerance to check against it.

jacobtomlinson reviewed (approved):
Much appreciated. I guess I was thinking about the explicit shape comparison in the issue, but asserting equality should do the same and more.
